### PR TITLE
Framework: Update npm packages to remove warnings from the console

### LINKED ISSUE #3107: Framework: Update dependencies to match the latest version of React
## Issue Overview

Raised by both @youknowriad and @aduth when upgrading React's version.

Reposting the comment from @aduth (https://github.com/WordPress/gutenberg/pull/3079#issuecomment-338214643):

> Noting that many of our dependencies define a React peer in the 15.x line, so warnings are logged when installing a new version of a package:
> 
> ```
> npm WARN react-autosize-textarea@0.4.8 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
> npm WARN react-autosize-textarea@0.4.8 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
> npm WARN react-datepicker@0.46.0 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
> npm WARN react-datepicker@0.46.0 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
> npm WARN react-slot-fill@1.0.1 requires a peer of react@^15.5.8 but none is installed. You must install peer dependencies yourself.
> npm WARN react-transition-group@1.2.0 requires a peer of react@^15.0.0 but none is installed. You must install peer dependencies yourself.
> npm WARN react-transition-group@1.2.0 requires a peer of react-dom@^15.0.0 but none is installed. You must install peer dependencies yourself.
> ```
> 
> I imagine at least some of these should have updated versions available which define React 16 as a valid peer.

I noticed also another package `babel-loader` which expects an older version of Webpack.

I did some research already and here are my findings:
* there are breaking changes introduced for `react-transition-group`. See [Migration Guide from v1 to v2](https://github.com/reactjs/react-transition-group/blob/master/Migration.md).
* `react-datepicker` has peer dependency updated in the master branch, but they haven't been published as a new version yet. I'm anticipating to see them soon in the upcoming `0.57.0` version. 

## Expected Behavior

No warnings from `npm install`.

## Current Behavior

```
npm WARN babel-loader@6.4.1 requires a peer of webpack@1 || 2 || ^2.1.0-beta || ^2.2.0-rc but none is installed. You must install peer dependencies yourself.
npm WARN react-autosize-textarea@0.4.8 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-autosize-textarea@0.4.8 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-datepicker@0.46.0 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-datepicker@0.46.0 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-slot-fill@1.0.1 requires a peer of react@^15.5.8 but none is installed. You must install peer dependencies yourself.
npm WARN react-transition-group@1.2.0 requires a peer of react@^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-transition-group@1.2.0 requires a peer of react-dom@^15.0.0 but none is installed. You must install peer dependencies yourself.
```

### PR DESCRIPTION
**Pull request by gziolo**
## Description

Fixes #3107. This PR updates a few npm packages to get rid of npm warnings:

```
npm WARN babel-loader@6.4.1 requires a peer of webpack@1 || 2 || ^2.1.0-beta || ^2.2.0-rc but none is installed. You must install peer dependencies yourself.
npm WARN react-autosize-textarea@0.4.8 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-autosize-textarea@0.4.8 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-datepicker@0.46.0 requires a peer of react@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-datepicker@0.46.0 requires a peer of react-dom@^0.14.0 || ^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-slot-fill@1.0.1 requires a peer of react@^15.5.8 but none is installed. You must install peer dependencies yourself.
npm WARN react-transition-group@1.2.0 requires a peer of react@^15.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-transition-group@1.2.0 requires a peer of react-dom@^15.0.0 but none is installed. You must install peer dependencies yourself.
```

There is still one warning logged on the console:

```
npm WARN ajv-keywords@2.1.0 requires a peer of ajv@>=5.0.0 but none is installed. You must install peer dependencies yourself.
```

It comes from the `enzyme` package which I updated to the latest version. I would wait until it's resolved on the `enzyme` side.

It also fixes new lint option for Jest:

> 30:27  warning  Use toHaveLength() instead  jest/prefer-to-have-length

## How Has This Been Tested?

`npm test` & `npm dev` still work.

Tested manually in the browser.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: blocks/block-icon/test/index.js [hunk 1]
```diff
@@ -40,14 +40,13 @@ describe( 'BlockIcon', () => {
 	} );
 
 	it( 'renders a component', () => {
+		class MyComponent extends Component {
```
Comment (gziolo) [inline blocks/block-icon/test/index.js]: Eslint had some issues with having this inlined ...

File: package.json [hunk 2]
```diff
@@ -34,14 +34,13 @@
...
+    "react": "16.2.0",
+    "react-autosize-textarea": "2.0.0",
     "react-click-outside": "2.3.1",
     "react-color": "2.13.4",
-    "react-datepicker": "0.46.0",
-    "react-dom": "16.0.0",
+    "react-datepicker": "0.61.0",
+    "react-dom": "16.2.0",
     "react-redux": "5.0.6",
-    "react-transition-group": "1.2.0",
```
Comment (gziolo) [inline package.json]: It looks like it is no longer used.

**Review by aduth — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -55,28 +54,28 @@
...
+    "babel-loader": "7.1.2",
     "babel-plugin-lodash": "3.2.11",
     "babel-plugin-transform-object-rest-spread": "6.26.0",
     "babel-plugin-transform-react-jsx": "6.24.1",
     "babel-plugin-transform-runtime": "6.23.0",
     "babel-preset-env": "1.6.0",
     "babel-traverse": "6.26.0",
     "check-node-version": "3.1.1",
-    "codecov": "2.3.0",
+    "codecov": "3.0.0",
```
Comment (mtias) [inline package.json]: Are we still using this?

Comment (gziolo) [reply package.json]: Yes, we display badge in Readme based on it.

**Review by ntwb — APPROVED** [review]

**Pull request merged**